### PR TITLE
Azure: Add CI pipeline for e2e test

### PR DESCRIPTION
**Pull request by kartikjoshi21**
Add workflow for e2e test. 

Steps: 
1. Build podvm image 
2. Setup cluster and test e2e
3. cleanup resources

### CODE REVIEW

File: .github/workflows/azure-e2e-test.yml [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        PKR_VAR_client_secret: ${{ secrets.AZURE_CLIENT_SECRET }}
+        PKR_VAR_tenant_id: ${{ secrets.AZURE_TENANT_ID }}
+        PKR_VAR_resource_group: ${{ secrets.AZURE_RESOURCE_GROUP }}
+        PKR_VAR_location: ${{ secrets.AZURE_REGION }}
+        PKR_VAR_az_image_name: ${{ env.PODVM_IMAGE_NAME }}
+        PKR_VAR_vm_size: ${{ env.VM_SIZE }}
+        PKR_VAR_ssh_username: ${{ env.SSH_USERNAME }}
+        PKR_VAR_az_gallery_name: ${{ env.PODVM_GALLERY_NAME }}
+        PKR_VAR_az_gallery_image_name: ${{ env.PODVM_IMAGE_DEF_NAME }}
+        PKR_VAR_az_gallery_image_version: ${{ env.PODVM_IMAGE_VERSION }}
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:77]: Since those env's are used only in this step, can we just set them here? like:

`PKR_VAR_vm_size: "Standard_D2as_v5"`

File: .github/workflows/azure-e2e-test.yml [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        RESOURCE_GROUP_NAME: ${{ secrets.AZURE_RESOURCE_GROUP }}
+        CLUSTER_NAME: "e2e-test-cluster"
+        LOCATION: ${{ secrets.AZURE_REGION }}
+        SSH_KEY_ID: "id_rsa.pub"
+        CLOUD_PROVIDER: "azure"
+        PODVM_IMAGE_NAME: ${{ env.PODVM_IMAGE_NAME }}
+        TEST_E2E_CREATE_RG: "no"
+      run: |
+        cat << EOF > /tmp/provision_azure.properties
+          AZURE_SUBSCRIPTION_ID="$AZURE_SUBSCRIPTION_ID"
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:115]: we should not need/have an indent here.

File: .github/workflows/azure-e2e-test.yml [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    - uses: ./.github/actions/login-to-azure
+      with:
+        client-id: ${{ secrets.AZURE_CLIENT_ID }}
+        client-secret: ${{ secrets.AZURE_CLIENT_SECRET }}
+        subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
+        tenant-id: ${{ secrets.AZURE_TENANT_ID }}
+
+    - name: Create provisioner file
+      env:
+        AZURE_SUBSCRIPTION_ID: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:102]: I guess there is no reason to put the variables into env's we can just use them in the cat << EOF directly.

File: .github/workflows/azure-e2e-test.yml [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+      - kartikjoshi21/ci-upstream
+
+jobs:
+  build-caa-image:
+    runs-on: ubuntu-latest
+    environment: build
+    steps:
+    - uses: actions/checkout@v2
+
+    - uses: ./.github/actions/login-to-azure
```
[the remark this answers is not in the document]
Comment (kartikjoshi21) [reply .github/workflows/azure-e2e-test.yml:23]: Need to change this way to use federated credentials. So will change it altogether. Suraj is working on how we can get it done without using client secret.

**Comment by bpradipt** [comment]
@kartikjoshi21 @sudharshanibm3 any possibility to get this in for 0.7.0 release ?

**Comment by surajssd** [comment]
@bpradipt this is blocked on: https://github.com/confidential-containers/cloud-api-adaptor/pull/1177

**Comment by bpradipt** [comment]
> @bpradipt this is blocked on: #1177

this is unblocked now.

File: .github/workflows/azure-e2e-test.yml [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        TEST_PROVISION: "yes"
+        CLOUD_PROVIDER: "azure"
+        BUILTIN_CLOUD_PROVIDERS: "azure"
+      run: | 
+        cat /tmp/provision_azure.properties
+        make test-e2e
+
+    - name: Remove provisioner file
+      run: |
+        rm -rf /tmp/provision_azure.properties
```
[the remark this answers is not in the document]
Comment (kartikjoshi21) [reply .github/workflows/azure-e2e-test.yml:128]: Kept it to make sure the file is removed

File: .github/workflows/azure-e2e-test.yml (hunk 1 above)
Comment (kartikjoshi21) [reply .github/workflows/azure-e2e-test.yml:60]: Later on we can plan to provide them as part of secrets.

File: .github/workflows/azure-e2e-test.yml [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+permissions:
+  id-token: write
+  contents: read
+
+env:
+  PODVM_GALLERY_NAME: "podvm_gallery3"
+  PODVM_IMAGE_DEF_NAME: "podvm_image3"
+  PODVM_IMAGE_VERSION: "0.0.3"
+  PODVM_IMAGE_NAME: "peerpod-image-0.0.1"
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:11]: nit: resource names look a bit arbitrary

File: .github/workflows/azure-e2e-test.yml [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  build-podvm-image:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+
+    - uses: ./.github/actions/setup-azure-build-env
+
+    - name: Clone kata repository
+      run: |
+        pushd ../
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:30]: I think each step starts fresh in the job's working dir, the PWD is not kept. so we don't need pushd/popd, maybe `git clone ... -b CCv0 ../kata-containers`?

File: .github/workflows/azure-e2e-test.yml [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        git clone https://github.com/kata-containers/kata-containers -b CCv0
+        popd
+
+    - name: Build binaries
+      env:
+        GOPATH: /home/runner/go
+      run: |
+        pushd azure/image
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:38]: we can just cd

File: .github/workflows/azure-e2e-test.yml (hunk 3 above)
Comment (mkulke) [reply .github/workflows/azure-e2e-test.yml:85]: can we move the envs into the ´cat EOF >>` expression directly to avoid duplication?

File: .github/workflows/azure-e2e-test.yml [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        LOCATION: ${{ secrets.AZURE_REGION }}
+        SSH_KEY_ID: "id_rsa.pub"
+        CLOUD_PROVIDER: "azure"
+        PODVM_IMAGE_NAME: ${{ env.PODVM_IMAGE_NAME }}
+        TEST_E2E_CREATE_RG: "no"
+      run: |
+        cat << EOF > /tmp/provision_azure.properties
+          AZURE_SUBSCRIPTION_ID="$AZURE_SUBSCRIPTION_ID"
+          AZURE_CLIENT_ID="$AZURE_CLIENT_ID"
+          AZURE_CLIENT_SECRET="$AZURE_CLIENT_SECRET"
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:99]: We don't need this anymore, right?

File: .github/workflows/azure-e2e-test.yml [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+          AZURE_TENANT_ID="$AZURE_TENANT_ID"
+          RESOURCE_GROUP_NAME="$RESOURCE_GROUP_NAME"
+          CLUSTER_NAME="$CLUSTER_NAME"
+          LOCATION="$LOCATION"
+          SSH_KEY_ID="$SSH_KEY_ID"
+          CLOUD_PROVIDER="$CLOUD_PROVIDER"
+          AZURE_IMAGE_ID="/subscriptions/$AZURE_SUBSCRIPTION_ID/resourceGroups/$RESOURCE_GROUP_NAME/providers/Microsoft.Compute/images/$PODVM_IMAGE_NAME"
+          IS_CI_MANAGED_CLUSTER="true"
+          AZURE_CLI_AUTH="true"
+          MANAGED_IDENTITY_NAME="ghactions_user3"
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:109]: nit: can we use a more descriptive name?

File: azure/image/ubuntu/azure-ubuntu.pkr.hcl [hunk 11]
```diff
@@ -21,6 +21,7 @@ source "azure-arm" "ubuntu" {
     image_name           = "${var.az_gallery_image_name}"
     image_version        = "${var.az_gallery_image_version}"
     storage_account_type = "Standard_LRS"
+    replication_regions  = ["eastus" , "westus"]
```
Comment (mkulke) [inline azure/image/ubuntu/azure-ubuntu.pkr.hcl:24]: do we need westus for an e2e test?

File: .github/workflows/azure-e2e-test.yml (hunk 6 above)
Comment (kartikjoshi21) [reply .github/workflows/azure-e2e-test.yml:11]: Planning to shift this to github secrets as well.

File: azure/image/ubuntu/azure-ubuntu.pkr.hcl (hunk 11 above)
Comment (kartikjoshi21) [reply azure/image/ubuntu/azure-ubuntu.pkr.hcl:24]: This is specific to packer version i guess, replication region is not required.

File: .github/workflows/azure-e2e-test.yml (hunk 10 above)
Comment (kartikjoshi21) [reply .github/workflows/azure-e2e-test.yml:109]: These envs name are from go env used in provisioner. If we need to change these we need to change provisioner code. I will make different change if these names need to be changed.

File: azure/image/ubuntu/azure-ubuntu.pkr.hcl (hunk 11 above)
Comment (mkulke) [reply azure/image/ubuntu/azure-ubuntu.pkr.hcl:24]: yeah, I mean we might be able to shorten the build time/czt resource use by not replicating it to westus

File: .github/workflows/azure-e2e-test.yml (hunk 6 above)
Comment (mkulke) [reply .github/workflows/azure-e2e-test.yml:11]: I wouldn't move this to secrets, those are not secret values. it will make the test harder to read and understand, the strings will be masked in the output and changes between runs are obscured.

File: .github/workflows/azure-e2e-test.yml [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        PODVM_IMAGE_NAME: ${{ env.PODVM_IMAGE_NAME }}
+        TEST_E2E_CREATE_RG: "no"
+      run: |
+        cat << EOF > /tmp/provision_azure.properties
+          AZURE_SUBSCRIPTION_ID=${{ secrets.AZURE_SUBSCRIPTION_ID }}
+          AZURE_CLIENT_ID=${{ secrets.AZURE_CLIENT_ID }}
+          AZURE_TENANT_ID=${{ secrets.AZURE_TENANT_ID }}
+          RESOURCE_GROUP_NAME=${{ secrets.AZURE_RESOURCE_GROUP }}
+          CLUSTER_NAME="$CLUSTER_NAME"
+          LOCATION=${{ secrets.AZURE_REGION }}
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:94]: note: secrets have to be put buffered in an env before using them in `run` blocks, otherwise you'll run into escaping issues due to special characters. it might work here, because of the EOF literal. but that should be standard practice:

```yaml
env:
 BLA="${{ secrets.bla }}
run: |
  cat << EOF > blub.properties
  BLA="$BLA"
  EOF
```  

also I would not store non-account related values like location and resource_group as secrets, but inline them here as env. that's maybe a matter of preference, but it will make it harder to debug, because those values will be hidden from the output and only users with access to secrets are able to understand what's going on.

File: .github/workflows/azure-e2e-test.yml [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+          SSH_KEY_ID="$SSH_KEY_ID"
+          CLOUD_PROVIDER="$CLOUD_PROVIDER"
+          AZURE_IMAGE_ID="/subscriptions/$AZURE_SUBSCRIPTION_ID/resourceGroups/$RESOURCE_GROUP_NAME/providers/Microsoft.Compute/images/$PODVM_IMAGE_NAME"
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:97]: `AZURE_SUBSCRIPTION_ID` is not env here right? (they should be though, since they are secrets). `RESOURCE_GROUP_NAME` can be an env too

File: .github/workflows/azure-e2e-test.yml [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+        PKR_VAR_vm_size: ${{ env.VM_SIZE }}
+        PKR_VAR_ssh_username: ${{ env.SSH_USERNAME }}
+        PKR_VAR_az_gallery_name: ${{ env.PODVM_GALLERY_NAME }}
+        PKR_VAR_az_gallery_image_name: ${{ env.PODVM_IMAGE_DEF_NAME }}
+        PKR_VAR_az_gallery_image_version: ${{ env.PODVM_IMAGE_VERSION }}
+        PKR_VAR_use_azure_cli_auth: "true"
+        GOPATH: /home/runner/go
+        PODVM_DISTRO: "ubuntu"
+      run: |
+        pushd azure/image
```
Comment (mkulke) [inline .github/workflows/azure-e2e-test.yml:62]: redundant push/pop

File: .github/actions/setup-azure-build-env/action.yml [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+     shell: bash
+
+   - name: Install Dependencies
+     run: |
+       rustup install 1.70.0
+       rustup default 1.70.0
+       rustup target add x86_64-unknown-linux-musl
+       sudo apt-get install musl-tools -y
+       sudo apt-get install -y libdevmapper-dev
+       sudo apt-get install -y libgpgme-dev
```
Comment (surajssd) [inline .github/actions/setup-azure-build-env/action.yml:26]:
```suggestion
       sudo apt-get install -y musl-tools libdevmapper-dev libgpgme-dev
```

File: .github/workflows/azure-e2e-test.yml [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+        PKR_VAR_resource_group: ${{ secrets.AZURE_RESOURCE_GROUP }}
+        PKR_VAR_location: ${{ secrets.AZURE_REGION }}
+        PKR_VAR_az_image_name: ${{ env.PODVM_IMAGE_NAME }}
+        PKR_VAR_vm_size: ${{ env.VM_SIZE }}
+        PKR_VAR_ssh_username: ${{ env.SSH_USERNAME }}
+        PKR_VAR_az_gallery_name: ${{ env.PODVM_GALLERY_NAME }}
+        PKR_VAR_az_gallery_image_name: ${{ env.PODVM_IMAGE_DEF_NAME }}
+        PKR_VAR_az_gallery_image_version: ${{ env.PODVM_IMAGE_VERSION }}
+        PKR_VAR_use_azure_cli_auth: "true"
+        GOPATH: /home/runner/go
```
Comment (surajssd) [inline .github/workflows/azure-e2e-test.yml:59]: This env var is already set in the `setup-azure-build-env` isn't it?

File: .github/workflows/azure-e2e-test.yml [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+    - uses: azure/login@v1
+      name: 'Az CLI login'
+      with:
+        client-id: ${{ secrets.AZURE_CLIENT_ID }}
+        subscription-id: ${{ secrets.AZURE_SUBSCRIPTION_ID }}
+        tenant-id: ${{ secrets.AZURE_TENANT_ID }}
+
+    - name: Create provisioner file
+      env:
+        CLUSTER_NAME: "e2e-test-cluster"
```
Comment (surajssd) [inline .github/workflows/azure-e2e-test.yml:82]: This name has to be auto generated based on PR number or something like that.

File: .github/workflows/azure-e2e-test.yml [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+        tenant-id: ${{ secrets.AZURE_TENANT_ID }}
+
+    - name: Remove podvm image
+      run: |
+        az sig image-version delete --gallery-name ${{ env.PODVM_GALLERY_NAME }} \
+                                    --gallery-image-definition ${{ env.PODVM_IMAGE_DEF_NAME }} \
+                                    --resource-group ${{ secrets.AZURE_RESOURCE_GROUP }} \
+                                    --gallery-image-version ${{ env.PODVM_IMAGE_VERSION }}
+        az image delete --resource-group ${{ secrets.AZURE_RESOURCE_GROUP }} \
+                        --name ${{ env.PODVM_IMAGE_NAME }}
```
Comment (surajssd) [inline .github/workflows/azure-e2e-test.yml:142]:
```suggestion
        az sig image-version delete --gallery-name ${{ env.PODVM_GALLERY_NAME }} \
          --gallery-image-definition ${{ env.PODVM_IMAGE_DEF_NAME }} \
          --resource-group ${{ secrets.AZURE_RESOURCE_GROUP }} \
          --gallery-image-version ${{ env.PODVM_IMAGE_VERSION }}
        az image delete --resource-group ${{ secrets.AZURE_RESOURCE_GROUP }} \
          --name ${{ env.PODVM_IMAGE_NAME }}
```

File: azure/image/ubuntu/variables.pkr.hcl [hunk 19]
```diff
@@ -113,4 +113,3 @@ variable "plan_publisher" {
   type    = string
   default = ""
 }
-
```
Comment (surajssd) [inline azure/image/ubuntu/variables.pkr.hcl:116]: Unrelated?

**Review by surajssd — COMMENTED** [review]
I will give it a try now.

File: .github/workflows/azure-e2e-test.yml (hunk 13 above)
Comment (kartikjoshi21) [reply .github/workflows/azure-e2e-test.yml:97]: we can directly use them as secrets. As per previous comment keeping them as env here will be redundant

File: .github/workflows/azure-e2e-test.yml (hunk 12 above)
Comment (kartikjoshi21) [reply .github/workflows/azure-e2e-test.yml:94]: OH, thats what i was doing earlier. i will change it to use envs.Thanks

**Comment by surajssd** [comment]
I have been iterating on this one as a test set up here: https://github.com/surajssd/caa-test1/pull/1

**Comment by kartikjoshi21** [comment]
> New set of blockers for this PR:
> 
> * [ ]  [azure-ci-static-infra: Output the managed identity name #1298](https://github.com/confidential-containers/cloud-api-adaptor/pull/1298)
> * [ ]  [azure-provisioner: Default to CVM instead of VM #1296](https://github.com/confidential-containers/cloud-api-adaptor/pull/1296)
> * [ ]  [azure-provisioner: Support for custom CAA images #1294](https://github.com/confidential-containers/cloud-api-adaptor/pull/1294)
> * [ ]  [azure-provisioner: Get RG when using CI managed RG #1292](https://github.com/confidential-containers/cloud-api-adaptor/pull/1292)

I have approved all of them, @mkulke  can you help to review them as well, Thanks

**Comment by mkulke** [comment]
@kartikjoshi21 i reviewed and merged all but #1294, but I think it's not a must-have for this PR. Can you address the remaining comments and add instructions on how to test this PR?

**Comment by surajssd** [comment]
## Instructions to test this PR

Follow instructions from here to do the set up: https://github.com/confidential-containers/cloud-api-adaptor/tree/staging/ci-infra/azure.

Here we will create Azure resources that will be needed to support the terraform infrastructure:

```bash
az logout

az login --use-device-code

export resource_group_name="REPLACE_ME"
export storage_account_name="REPLACE_ME"
export container_name="terraform-state"
export location="eastus"
```

Create azure resources:

```bash
az group create \
    --name $resource_group_name \
    --location $location

az storage account create \
    --name $storage_account_name \
    --resource-group $resource_group_name \
    --location $location \
    --sku Standard_LRS \
    --kind StorageV2

az storage container create \
    --name $container_name \
    --account-name $storage_account_name
```

Clone a fresh copy of the CAA:

```bash
cd $(mktemp -d)
git clone git@github.com:confidential-containers/cloud-api-adaptor.git

cd cloud-api-adaptor

git remote remove origin
git remote add upstream git@github.com:confidential-containers/cloud-api-adaptor.git
gh repo set-default confidential-containers/cloud-api-adaptor
```

Let's create a new repository to use as a playground for this PR:

```bash
export version="1"
export reponame="caa-test${version}"
gh repo create "${reponame}" --public --source . --remote=origin
git push -u origin main
```

Checkout this PR's code:

```bash
gh pr checkout 932
```

Modify two fields from `providers.tf` to match the env vars exported at the beggining viz. `resource_group_name` and `storage_account_name`:

```bash
cd ci-infra/azure
vim providers.tf
```

Populate the terraform variables file:

```bash
export username="REPLACE_ME"
export org_full_name="${username}/${reponame}"

cat <<EOF >terraform.tfvars
ver                = "${version}"
container_registry = "surajd"
gh_repo            = "${org_full_name}"
EOF
```

Finally create the static infra using terrform:

```bash
terraform init
terraform apply
```

Make a note of the output of the above command, above env var should be used to create secret on the newly created GH repo. To create those secrets from command line, populate the following values accordingly:

```bash
gh secret set ACR_URL --app actions -R "${org_full_name}" --body ""
gh secret set AZURE_CLIENT_ID --app actions -R "${org_full_name}" --body ""
gh secret set AZURE_PODVM_GALLERY_NAME --app actions -R "${org_full_name}" --body ""
gh secret set AZURE_PODVM_IMAGE_DEF_NAME --app actions -R "${org_full_name}" --body ""
gh secret set AZURE_REGION --app actions -R "${org_full_name}" --body ""
gh secret set AZURE_RESOURCE_GROUP --app actions -R "${org_full_name}" --body ""
gh secret set AZURE_SUBSCRIPTION_ID --app actions -R "${org_full_name}" --body ""
gh secret set AZURE_TENANT_ID --app actions -R "${org_full_name}" --body ""
gh secret set AZURE_MANAGED_IDENTITY_NAME --app actions -R "${org_full_name}" --body ""
```

Commit the terraform changes to trigger a CI run:

```bash
git add .
git commit -m "update terraform"

git push -u origin $(git branch --show-current)
```

**Comment by surajssd** [comment]
So the secrets are there on this repo, but the pipeline cannot access it because they are not on the Kartik's fork. So need to figure out how to get these secrets available to the PRs.

**Comment by kartikjoshi21** [comment]
To access secrets in forked repo PRs. [This](https://iterative.ai/blog/testing-external-contributions-using-github-actions-secrets) is the available option with github CI. 

Although for this the CI PR should be merged first then it will run on approval on PRs which are submitted after that. So if we need to make any changes to CI workflow we will have to test and run it in public forked repo and provide reference here for test results. 
cc: @surajssd @mkulke  @bpradipt

**Comment by bpradipt** [comment]
> To access secrets in forked repo PRs. [This](https://iterative.ai/blog/testing-external-contributions-using-github-actions-secrets) is the available option with github CI.
> 
> Although for this the CI PR should be merged first then it will run on approval on PRs which are submitted after that. So if we need to make any changes to CI workflow we will have to test and run it in public forked repo and provide reference here for test results. cc: @surajssd @mkulke @bpradipt

Can you also please add a description in the https://github.com/confidential-containers/cloud-api-adaptor/tree/main/ci-infra directory on the workflow and what needs to be done when updating the specific CI workflow. This will be a good reference.

**Comment by bpradipt** [comment]
@kartikjoshi21 I have created the `external` environment. Please add it to the docs.

**Review by bpradipt — APPROVED** [review]
/lgtm
Once the doc is added, this should be good to merge

**Comment by kartikjoshi21** [comment]
> /lgtm Once the doc is added, this should be good to merge

Added doc for the same. Thanks @bpradipt

**Review by surajssd — APPROVED** [review]
